runtime-sdk/src/module.rs

```diff
@@ -8,137 +8,113 @@ use oasis_core_runtime::common::cbor;
 
 use crate::{
     context::{DispatchContext, TxContext},
-    error, event, modules, storage,
+    error, event, modules,
+    runtime::Runtime,
+    storage,
     storage::Store,
     types::{
         message::MessageEvent,
         transaction::{CallResult, Transaction, UnverifiedTransaction},
     },
 };
 
-/// Metadata of a callable method.
-pub struct CallableMethodInfo {
-    /// Method name.
-    pub name: &'static str,
-
-    /// Method handler function.
-    pub handler: fn(&CallableMethodInfo, &mut TxContext<'_, '_>, cbor::Value) -> CallResult,
-}
-
-/// Metadata of a query method.
-pub struct QueryMethodInfo {
-    /// Method name.
-    pub name: &'static str,
-
-    /// Method handler function.
-    pub handler: fn(
-        &QueryMethodInfo,
-        &mut DispatchContext<'_>,
-        cbor::Value,
-    ) -> Result<cbor::Value, error::RuntimeError>,
-}
-
-/// Registry of methods exposed by the modules.
-pub struct MethodRegistry {
-    callable_methods: BTreeMap<&'static str, CallableMethodInfo>,
-    query_methods: BTreeMap<&'static str, QueryMethodInfo>,
+pub enum DispatchResult<B, R> {
+    Handled(R),
+    Unhandled(B),
 }
 
-impl MethodRegistry {
-    /// Create a new method registry.
-    #[allow(clippy::new_without_default)]
-    pub fn new() -> Self {
-        Self {
-            callable_methods: BTreeMap::new(),
-            query_methods: BTreeMap::new(),
+impl<B, R> DispatchResult<B, R> {
+    pub fn ok_or<E>(self, err: E) -> Result<R, E> {
+        match self {
+            DispatchResult::Handled(result) => Ok(result),
+            DispatchResult::Unhandled(_) => Err(err),
         }
     }
+}
 
-    /// Register a new callable method.
-    ///
-    /// # Panics
-    ///
-    /// This method will panic in case a method with the same name is already registered.
-    pub fn register_callable(&mut self, mi: CallableMethodInfo) {
-        if self.callable_methods.contains_key(mi.name) {
-            panic!("callable method already exists: {}", mi.name);
-        }
-        self.callable_methods.insert(mi.name, mi);
-    }
+#[derive(Clone, Debug)]
+pub struct MessageResult {
+    pub event: MessageEvent,
+    pub context: cbor::Value,
+}
 
-    /// Looks up a previously registered callable method.
-    pub fn lookup_callable(&self, name: &str) -> Option<&CallableMethodInfo> {
-        self.callable_methods.get(name)
+pub trait Dispatcher {
```

Comment:
could we consider a different name, Dispatcher -> MethodHandler
- for some kind of consistency with AuthHandler/BlockHandler/etc.
- to make the codebase easier to navigate by not duplicating dispatcher::Dispatcher's name

Reply:
Agreed.